### ISSUE TITLE
Should `transcode_to` for well known types be using the descriptors in `DescriptorPool::global`? 

### ISSUE THREAD

**Issue by slinkydeveloper** [andrewhickman/prost-reflect#49]
tl:dr: due to a bug in the library generating the descriptor for wkt types, prost-reflect ser/de support doesn't work correctly.

## Background

I have a system where I'm using descriptors from `ts-proto`, and I'm then parsing those descriptors using `prost-reflect` to support JSON transcoding.

When trying to use `google.protobuf.Value` in a message definition, I stumbled on an interesting issue: `google.protobuf.Value` has only one field, defined as `oneof`, called `kind`. This field has index 0. Due to the fact that ts-proto doesn't correctly encode [proto2 optional fields](https://github.com/stephenh/ts-proto/issues/139#issuecomment-980465063) and the explicit presence of proto2, the `oneof_index` of the field descriptors is not parsed, thus generating an invalid descriptor for this type:

```
MessageDescriptor {
    name: "Value",
    full_name: "google.protobuf.Value",
    is_map_entry: false,
    fields: [
        FieldDescriptor {
            name: "null_value",
            full_name: "google.protobuf.Value.null_value",
            json_name: "nullValue",
            number: 1,
            kind: google.protobuf.NullValue,
            cardinality: Optional,
            containing_oneof: None,
            default_value: None,
            is_group: false,
            is_list: false,
            is_map: false,
            is_packed: false,
            supports_presence: false,
        },
        FieldDescriptor {
            name: "number_value",
            full_name: "google.protobuf.Value.number_value",
            json_name: "numberValue",
            number: 2,
            kind: double,
            cardinality: Optional,
            containing_oneof: None,
            default_value: None,
            is_group: false,
            is_list: false,
            is_map: false,
            is_packed: false,
            supports_presence: false,
        },
        FieldDescriptor {
            name: "string_value",
            full_name: "google.protobuf.Value.string_value",
            json_name: "stringValue",
            number: 3,
            kind: string,
            cardinality: Optional,
            containing_oneof: None,
            default_value: None,
            is_group: false,
            is_list: false,
            is_map: false,
            is_packed: false,
            supports_presence: false,
        },
        FieldDescriptor {
            name: "bool_value",
            full_name: "google.protobuf.Value.bool_value",
            json_name: "boolValue",
            number: 4,
            kind: bool,
            cardinality: Optional,
            containing_oneof: None,
            default_value: None,
            is_group: false,
            is_list: false,
            is_map: false,
            is_packed: false,
            supports_presence: false,
        },
        FieldDescriptor {
            name: "struct_value",
            full_name: "google.protobuf.Value.struct_value",
            json_name: "structValue",
            number: 5,
            kind: google.protobuf.Struct,
            cardinality: Optional,
            containing_oneof: None,
            default_value: None,
            is_group: false,
            is_list: false,
            is_map: false,
            is_packed: false,
            supports_presence: true,
        },
        FieldDescriptor {
            name: "list_value",
            full_name: "google.protobuf.Value.list_value",
            json_name: "listValue",
            number: 6,
            kind: google.protobuf.ListValue,
            cardinality: Optional,
            containing_oneof: None,
            default_value: None,
            is_group: false,
            is_list: false,
            is_map: false,
            is_packed: false,
            supports_presence: true,
        },
    ],
    oneofs: [
        OneofDescriptor {
            name: "kind",
            full_name: "google.protobuf.Value.kind",
            fields: [],
        },
    ],
},
```

Because of this bad descriptor, when trying to serialize my type (that includes the well known type `google.protobuf.Value`), `transcode_to` returns a wrong result, leading to a wrong serialized JSON. For example, by implementing a roundtrip test (deserialize then serialize) for `google.protobuf.Value` with this JSON:

```json
{ "a": false, "b": 0, "c": "", "d": true, "e": 1, "f": "hello"}
``` 

When using the `DescriptorPool::global`, or any descriptor pool generated using `prost_build`, I get the same output, while with my "bad" descriptor pool, I get the following:

```json
{"a":null,"b":null,"c":null,"d":true,"e":1.0,"f":"hello"}
```

## Question

In general what I found surprising (and I might be wrong here) is that for these well known types my descriptor is affecting the transcoding logic. The serde `Serializer`/`Deserializer` traits are effectively hardcoded for `google.protobuf.Value`, and the descriptor is already loaded in `DescriptorPool::global`, but nevertheless my descriptor is used for `transcode_to`. Shouldn't we use the descriptors in `DescriptorPool::global` every time a `DynamicMessage` with a well known type is created?

Perhaps a bit tangential question, but it could help me mitigate this issue: would it be possible to create a descriptor pool with my descriptor files, and then override the descriptors for well known types with the ones in `DescriptorPool::global`?

**Comment by andrewhickman** [comment]
To work around your issue, I think you could use `DescriptorPool::global` to get the well-known types, and then add your own descriptors on top:
```rust
let mut pool = DescriptorPool::global();
pool.add_file_descriptor_set(file_descriptor_set);
```
your copies of the well-known file descriptors should be skipped since a file by that name already exists in the pool

The reason for using the user-provided descriptor by default is to avoid issues with different versions of the well-known file descriptors.

**Comment by slinkydeveloper** [comment]
> pool.add_file_descriptor_set(file_descriptor_set);

But this requires me to parse the file_descriptor_set using prost types, right? Wouldn't I lose extensions with that?

**Comment by andrewhickman** [comment]
You could use  [decode_file_descriptor_set](https://docs.rs/prost-reflect/latest/prost_reflect/struct.DescriptorPool.html#method.decode_file_descriptor_set) instead if you want to preserve extension options.

**Comment by slinkydeveloper** [comment]
That's pretty much what I'm doing, but at the cost of copy-pasting the descriptors defined in this module:

```rust
fn patch_built_in_descriptors(mut files: Bytes) -> Result<Bytes, ServiceDiscoveryError> {
    // prost_reflect_types is copy pasted from https://github.com/andrewhickman/prost-reflect/blob/8fc52d4c587117ca68c29bf27034d16803d97803/prost-reflect/src/descriptor/types.rs
    // because due to the package private visibility  
    let mut files = prost_reflect_types::FileDescriptorSet::decode(&mut files)?;

    // Let's patch the file google/protobuf/struct.proto
    for file in &mut files.file {
        if file.name() == "google/protobuf/struct.proto" {
            // Let's take the descriptor we need from the DescriptorPool::global()
            let file_desc = DescriptorPool::global()
                .get_file_by_name("google/protobuf/struct.proto")
                .expect("The global descriptor pool must contain struct.proto")
                .encode_to_vec();

            // Let's apply it
            *file = prost_reflect_types::FileDescriptorProto::decode(&*file_desc)
                .expect("This deserialization should not fail!");
        }
    }

    Ok(Bytes::from(files.encode_to_vec()))
}
```
